### ISSUE TITLE
Istio and kiam/kube2iam together

### ISSUE THREAD

**Issue by kgoralski** [istio/istio#9297]
Hi! I have a question. Can I use tools like kiam/kube2iam or GoogleCloudPlatform/k8s-metadata-proxy that integrates AWS IAM or GCP (Metadata API) with k8s.... together with istio? Can I install it and it should just work or I need to configure it somehow?

Installed kiam (without istio and sidecar) and after that installed istio and everything seems to work fine, but actually I'm not sure how everything works and how traffic forwarding looks like. ;)

anything related:
Comparison of iam4kube, kiam, kube2iam etc. https://docs.google.com/document/d/1rn-v2TNH9k4Oz-VuaueP77ANE5p-5Ua89obK2JaArfg/edit
Istio and kube2iam https://groups.google.com/forum/#!topic/istio-users/3-fp2JPb2dQ
https://medium.com/@pingles/kiam-iterating-for-security-and-reliability-5e793ab93ec3
https://github.com/GoogleCloudPlatform/k8s-metadata-proxy
https://github.com/envoyproxy/envoy/issues/170#issuecomment-348576121
https://github.com/uswitch/kiam/issues/13
https://github.com/jtblin/kube2iam/issues/79
https://istio.io/docs/concepts/security/#istio-identity
https://stackoverflow.com/questions/52732417/integration-istio-with-aws-iam

**Comment by ashlineldridge** [comment]
@kgoralski I tried it the other way around (installed Istio first and then Kiam) and it didn't work for me. Did you verify that Istio routing works as expected within the mesh?

The agent/server model seems like it would be unnecessary with Istio given that Envoy could be used to provide load balancing and mTLS between client services and the Kiam service.

@pingles Would it be possible to deploy the Kiam server (without modification?) as an Istio-enabled service and then create a mesh routing rule to route all `169.254.169.254` bound traffic to the server? Or is something special taking place within the Kiam agent that would make this untenable?

**Comment by pingles** [comment]
You'd want to route `169.254.169.254` traffic to the agent process (which expects to run on each node and acts as the HTTP metadata service). That will make gRPC calls to the server process (it uses the client-side load-balancing currently but istio could displace this.

I agree @ashlineldridge: much of Kiam's agent/server topology should be unnecessary in an istio world: the mesh would provide the retry/failure handling that is encoded within the agent and gRPC load-balancing lib and could manage the mTLS. I wrote a [medium post](https://medium.com/@pingles/kiam-iterating-for-security-and-reliability-5e793ab93ec3) about how we ended up with Kiam in the shape it is. 

There should be a way we could change the servers/agents to be simpler and not break existing users when run on top of a mesh like Istio that provides equivalent capabilities. I'd definitely support such changes. I'm sure that we'll end up running Istio but it's unlikely to be until next year so I don't think we could lead on it until then though 😞

**Comment by pingles** [comment]
You'd want to route `169.254.169.254` traffic to the agent process (which expects to run on each node and acts as the HTTP metadata service). That will make gRPC calls to the server process (it uses the client-side load-balancing currently but istio could displace this.

I agree @ashlineldridge: much of Kiam's agent/server topology should be unnecessary in an istio world: the mesh would provide the retry/failure handling that is encoded within the agent and gRPC load-balancing lib and could manage the mTLS. I wrote a [medium post](https://medium.com/@pingles/kiam-iterating-for-security-and-reliability-5e793ab93ec3) about how we ended up with Kiam in the shape it is. 

There should be a way we could change the servers/agents to be simpler and not break existing users when run on top of a mesh like Istio that provides equivalent capabilities. I'd definitely support such changes. I'm sure that we'll end up running Istio but it's unlikely to be until next year so I don't think we could lead on it until then though 😞

**Comment by pingles** [comment]
You'd want to route `169.254.169.254` traffic to the agent process (which expects to run on each node and acts as the HTTP metadata service). That will make gRPC calls to the server process (it uses the client-side load-balancing currently but istio could displace this.

I agree @ashlineldridge: much of Kiam's agent/server topology should be unnecessary in an istio world: the mesh would provide the retry/failure handling that is encoded within the agent and gRPC load-balancing lib and could manage the mTLS. I wrote a [medium post](https://medium.com/@pingles/kiam-iterating-for-security-and-reliability-5e793ab93ec3) about how we ended up with Kiam in the shape it is. 

There should be a way we could change the servers/agents to be simpler and not break existing users when run on top of a mesh like Istio that provides equivalent capabilities. I'd definitely support such changes. I'm sure that we'll end up running Istio but it's unlikely to be until next year so I don't think we could lead on it until then though 😞

**Comment by pingles** [comment]
You'd want to route `169.254.169.254` traffic to the agent process (which expects to run on each node and acts as the HTTP metadata service). That will make gRPC calls to the server process (it uses the client-side load-balancing currently but istio could displace this.

I agree @ashlineldridge: much of Kiam's agent/server topology should be unnecessary in an istio world: the mesh would provide the retry/failure handling that is encoded within the agent and gRPC load-balancing lib and could manage the mTLS. I wrote a [medium post](https://medium.com/@pingles/kiam-iterating-for-security-and-reliability-5e793ab93ec3) about how we ended up with Kiam in the shape it is. 

There should be a way we could change the servers/agents to be simpler and not break existing users when run on top of a mesh like Istio that provides equivalent capabilities. I'd definitely support such changes. I'm sure that we'll end up running Istio but it's unlikely to be until next year so I don't think we could lead on it until then though 😞

**Comment by pingles** [comment]
You'd want to route `169.254.169.254` traffic to the agent process (which expects to run on each node and acts as the HTTP metadata service). That will make gRPC calls to the server process (it uses the client-side load-balancing currently but istio could displace this.

I agree @ashlineldridge: much of Kiam's agent/server topology should be unnecessary in an istio world: the mesh would provide the retry/failure handling that is encoded within the agent and gRPC load-balancing lib and could manage the mTLS. I wrote a [medium post](https://medium.com/@pingles/kiam-iterating-for-security-and-reliability-5e793ab93ec3) about how we ended up with Kiam in the shape it is. 

There should be a way we could change the servers/agents to be simpler and not break existing users when run on top of a mesh like Istio that provides equivalent capabilities. I'd definitely support such changes. I'm sure that we'll end up running Istio but it's unlikely to be until next year so I don't think we could lead on it until then though 😞

**Comment by geota** [comment]
We are looking to integrate istio + kiam and would be happy to help contribute to changes needed in the projects to support them working together. 

Other things we are looking to play nice together is calico in policy engine only mode + kiam.

https://www.projectcalico.org/using-network-policy-in-concert-with-istio/

**Comment by nemo83** [comment]
The istio docs say kiam / kube2iam should not be required at all: https://preliminary.istio.io/docs/concepts/security/#istio-identity
Unfortunately thou, I cannot find any further doc on how really implement it :-(

**Comment by eigokor** [comment]
Looks like found one solution

adding proxy.excludeIPRanges: "169.254.169.254/32"

**Comment by crhuber** [comment]
I can confirm we also had issues with Kiam and Istio. 
We have a python application that connects to S3 on k8s & istio. On attempts to connect to s3 the application would get an access denied message even though KIAM and AWS roles seemed to be setup correctly. Upon disabling the sidecar, s3 access worked as expected. 

However, when we started a second processes of our application within the same container/pod with Istio enabled, s3 access worked fine. So, it appears to be some kind of race condition between the istio-proxy sidecar and the application container. 

We were able to work around this issue by adding 

~~~
adding proxy.excludeIPRanges: "169.254.169.254/32"
~~~

to our istio helm values
